### ISSUE TITLE
Memory leak is possible for UDF array arguments

### ISSUE THREAD

**Issue by dmitry-lipetsk** [FirebirdSQL/firebird#7817]
FB v4.0.4.2988

If this line throws
https://github.com/FirebirdSQL/firebird/blob/9591891e906df94d66589063fcab3b0333445607/src/jrd/fun.epp#L1018
you do not release the allocated memory in 'data'
https://github.com/FirebirdSQL/firebird/blob/9591891e906df94d66589063fcab3b0333445607/src/jrd/fun.epp#L1005

Bricks for reproducing this problem:

```sql
DECLARE EXTERNAL FUNCTION MY_UDF__GET_DIM
 INTEGER BY SCALAR_ARRAY
RETURNS INTEGER BY VALUE
ENTRY_POINT 'fn_get_dim__i4__sa'
MODULE_NAME 'lcpi.test.udf.01.dll'

select MY_UDF__GET_DIM(cast('12311111111111111111111111111111111111111111111111111111111111111111111111111111111' as blob)) from rdb$database
```

![image](https://github.com/FirebirdSQL/firebird/assets/4594554/fe32d8b2-29e0-434d-a599-fca1659f2bce)

Memory releases after disconnect.

**Comment by dyemanov** [comment]
Please verify the fix and I will port it to other versions.

**Comment by dmitry-lipetsk** [comment]
Good time of day.

Fix looks ok.

![image](https://github.com/FirebirdSQL/firebird/assets/4594554/1f9b8ed2-b82c-4eec-90b6-4a3751f73c24)

![image](https://github.com/FirebirdSQL/firebird/assets/4594554/c461dfd8-c982-4e7d-9879-3c446c8d4fb9)

Another simple tests work fine, too.

**Source code of test UDF:**

``` c++
#include "source/fn.h"
#include <cstddef>

////////////////////////////////////////////////////////////////////////////////

using SSHORT=std::int16_t;
using SLONG =std::int32_t;
using SCHAR =char;

////////////////////////////////////////////////////////////////////////////////

typedef struct dsc
{
 UCHAR  dsc_dtype;
 SCHAR  dsc_scale;
 USHORT dsc_length;
 SSHORT dsc_sub_type;
 USHORT dsc_flags;
 UCHAR* dsc_address; // Used either as offset in a message or as a pointer
} DSC;

struct scalar_array_desc
{
 DSC   sad_desc;
 SLONG sad_dimensions;
 
 struct sad_repeat
 {
  SLONG sad_lower;
  SLONG sad_upper;
 } sad_rpt[1];
};//struct scalar_array_desc

////////////////////////////////////////////////////////////////////////////////

std::int32_t FBUDF_API fn_get_dim__i4__sa(const scalar_array_desc* const descr)
{
 if(descr->sad_desc.dsc_address==nullptr)
  return 0;

 return descr->sad_dimensions;
}//fn_get_dim__i4__sa

////////////////////////////////////////////////////////////////////////////////

```

**Comment by onlylunix** [comment]
Will there be a fix for version 3.0.11?

**Comment by dyemanov** [comment]
> Will there be a fix for version 3.0.11?

Backported.
